src/org/labkey/remoteapi/Command.java

```diff
@@ -107,12 +108,22 @@ public String getActionName()
         return _actionName;
     }
 
+    /**
+     * Makes an immutable copy of the parameter map used for building the URL available to callers. Typically used for
+     * logging and testing.
+     * @return An immutable map of parameters used when building the URL.
+     */
+    public Map<String, Object> getParameters()
+    {
+        return Collections.unmodifiableMap(createParameterMap());
+    }
+
     /**
      * Returns a new, mutable parameter map. Derived classes will typically override this
      * method to put values passed to specialized setter methods into the map.
      * @return The parameter map to use when building the URL.
      */
-    public Map<String, Object> getParameters()
+    protected Map<String, Object> createParameterMap()
```

Comment:
Having two methods to get parameters seems like a recipe for confusion as to which one should be overridden by subclasses. What if we made `getParameters` final?

Reply:
Yes, I meant to do that. Good catch.